In [2]:
import pandas as pd

df = pd.read_csv('churnguard_data.csv')
df

,customerID,gender,SeniorCitizen,tenure,PhoneService,InternetService,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,CUST-0032,Male,0,21.0,YES,Fiber optic,Month-to-month,No,Credit card,29.73,600.01,Yes
1,CUST-0110,Male,0,55.0,YES,Fiber optic,Two year,Yes,Bank transfer,46.32,2515.48,No
2,CUST-0137,Female,1,46.0,Yes,Fiber optic,Month-to-month,No,Mailed check,87.06,4153.97,Yes
3,CUST-0089,Female,1,63.0,Yes,Fiber optic,Month-to-month,YES,Mailed check,56.97,3641.3,Yes
4,CUST-0919,Female,0,8.0,Yes,DSl,month to month,No,Electronic check,39.69,309.79,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...
1025,CUST-0088,Female,0,26.0,Yes,DSL,One year,Yes,Electronic check,51.68,1369.73,No
1026,CUST-0331,Male,0,49.0,Yes,Fiber optic,Two year,Yes,Electronic check,85.39,4305.4,No
1027,CUST-0467,Male,1,69.0,Yes,DSl,Two year,Yes,Electronic check,95.00,6689.62,No
1028,CUST-0122,Female,1,3.0,Yes,No,Two year,Yes,Electronic check,25.08,77.2,No


In [3]:
df.shape

(1030, 12)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        1030 non-null   object 
 1   gender            1030 non-null   object 
 2   SeniorCitizen     1030 non-null   int64  
 3   tenure            980 non-null    float64
 4   PhoneService      1030 non-null   object 
 5   InternetService   1014 non-null   object 
 6   Contract          1030 non-null   object 
 7   PaperlessBilling  1030 non-null   object 
 8   PaymentMethod     1030 non-null   object 
 9   MonthlyCharges    955 non-null    float64
 10  TotalCharges      966 non-null    object 
 11  Churn             1030 non-null   object 
dtypes: float64(2), int64(1), object(9)
memory usage: 96.7+ KB


In [8]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,1030.000000,980.000000,955.000000
mean,0.166990,35.632653,70.539738
std,0.373148,20.427948,80.661140
min,0.000000,-5.000000,0.010000
25%,0.000000,18.000000,42.640000
50%,0.000000,36.000000,64.970000
75%,0.000000,53.000000,89.165000
max,1.000000,71.000000,1500.000000


In [9]:
df.isna().sum()

customerID           0
gender               0
SeniorCitizen        0
tenure              50
PhoneService         0
InternetService     16
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges      75
TotalCharges        64
Churn                0
dtype: int64

In [15]:
df.duplicated().sum()

np.int64(30)

In [17]:
df['Churn'].value_counts()

Churn
No     598
Yes    278
NO      38
no      38
nO      27
yEs     20
YES     20
yes     11
Name: count, dtype: int64

In [19]:
df['Contract'].unique()

array(['Month-to-month', 'Two year', 'month to month', 'One year',
       'month-to-month', 'Monthly', 'Two Year', '1 year', '2 year',
       'two year', 'One Year', 'one year'], dtype=object)

# Task 1 — Load and Explore Dataset (ChurnGuard)

## Dataset Overview
- **File:** churnguard_data.csv
- **Shape:** 1030 rows, 12 columns
- **Goal:** Identify customers likely to cancel subscription (Churn prediction)

## Column Summary
| Column | Type | Description |
|--------|------|-------------|
| customerID | object | Unique customer ID |
| gender | object | Customer gender |
| SeniorCitizen | int64 | 1=senior, 0=not |
| tenure | float64 | Months with company |
| PhoneService | object | Has phone service |
| InternetService | object | Type of internet |
| Contract | object | Contract type |
| PaperlessBilling | object | Paperless billing |
| PaymentMethod | object | Payment method |
| MonthlyCharges | float64 | Monthly bill |
| TotalCharges | object | Total billed (stored as TEXT!) |
| Churn | object | Target — left or not |

## Data Quality Issues Found

### 1. Missing Values
| Column | Missing Count |
|--------|--------------|
| tenure | 50 |
| MonthlyCharges | 75 |
| TotalCharges | 64 |
| InternetService | 16 |

### 2. Duplicate Rows
30 duplicate rows found → need to remove!
### 3. Churn Column — Inconsistent Casing
No, NO, no, nO   → all mean "No"
Yes, YES, yes, yEs → all mean "Yes"
Should be only: Yes / No

### 4. Contract Column — Typos and Inconsistent Formats
12 unique values found → should be only 3!
Month-to-month = month to month = month-to-month = Monthly
One year       = One Year       = one year       = 1 year
Two year       = Two Year       = two year       = 2 year

### 5. tenure — Impossible Values
min = -5 (negative tenure is impossible!)
→ needs to be handled

### 6. MonthlyCharges — Possible Outlier
max = 1500 (very high compared to mean=70)
→ needs investigation

### 7. TotalCharges — Wrong Data Type
Stored as object (text) instead of float
→ needs conversion to numeric

## Code Used

```python
import pandas as pd

df = pd.read_csv('churnguard_data.csv')

# Shape
print("Shape:", df.shape)

# First 5 rows
print(df.head())

# Column types and info
df.info()

# Missing values
print(df.isnull().sum())

# Duplicates
print(df.duplicated().sum())

# Churn inconsistency
print(df['Churn'].value_counts())

# Contract typos
print(df['Contract'].unique())
```

## Key Takeaways
- Real world data is always messy!
- Always explore before cleaning
- Check: shape, dtypes, missing values, duplicates, unique values
- TotalCharges stored as text → classic real world issue
- Inconsistent casing and typos are very common in collected data